sdk/metric/provider.go

```diff
@@ -0,0 +1,96 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric // import "go.opentelemetry.io/otel/sdk/metric"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/metric"
+)
+
+// MeterProvider handles the creation and coordination of Meters. All Meters
+// created by a MeterProvider will be associated with the same Resource, have
+// the same Views applied to them, and have their produced metric telemetry
+// passed to the configured Readers.
+type MeterProvider struct {
+	// TODO (#2820): implement.
+}
+
+// Compile-time check MeterProvider implements metric.MeterProvider.
+var _ metric.MeterProvider = (*MeterProvider)(nil)
+
+// NewMeterProvider returns a new and configured MeterProvider.
+//
+// By default, the returned MeterProvider is configured with the default
+// Resource and no Readers.
```

Comment:
Because of our discussion today, should we mention readers?  

If we create readers that encapsulate MeterProviders, then no MeterProvider will ever be configured with a Reader.

Reply:
```suggestion
// Resource.
```

Reply:
I don't understand the goal of "readers that encapsulate MeterProviders". 

The SDK spec clearly calls out a MeterProvider.ForceFlush method that will arrange to flush each reader/exporter in addition to Shutdown.

The way I understand it, the piece of code responsible for creating and using an SDK is also the piece of code that's responsible for flush and shutdown. Those signals -- flush and shutdown -- come from the user to the SDK to the reader/exporter. The other signal -- collect -- comes from the reader/exporter to the SDK.

I'm not inclined to allow Readers (and their Views) to be added dynamically, it just seems like unnecessary complexity that's not part of the SDK specification. Thus, I see no reason to create a Reader-less MeterProvider, it would be equivalent to a No-op meter.

Reply:
> The way I understand it, the piece of code responsible for creating and using an SDK is also the piece of code that's responsible for flush and shutdown. Those signals -- flush and shutdown -- come from the user to the SDK to the reader/exporter.

Right, I don't see this changing. The readers would still need to have their flush/shutdown methods registered with the MeterProvider they are passed when they are created.

> The other signal -- collect -- comes from the reader/exporter to the SDK.

Exactly. This is where the telemetry control flow issue arises and where the proposal to pass a singular MeterProvider to a Reader comes from.

> I'm not inclined to allow Readers (and their Views) to be added dynamically, it just seems like unnecessary complexity that's not part of the SDK specification.

I don't see how registering a shutdown/flush method with a MeterProvider is more complex than creating and updating reader/view mappings in a MeterProvider and a view Controller? Registering two methods that need to be called by the MeterProvider seems simpler than dynamically filtering what views to apply to the caller of a produce function every collection cycle.

>  Thus, I see no reason to create a Reader-less MeterProvider, it would be equivalent to a No-op meter.

Agreed. If the MeterProvider was never passed to a reader it would be a no-op in the case where the MeterProvider is passed to a Reader as well.

Reply:
These sentences make sense when you are operating in a world where the MeterProvider contains within it some number of Readers, ie. WithReader, which you can flush.  But, if we do go down the road of Readers are created with a MeterProvider, ie. reader.New(MeterProvider...), the Meterprovider doesn't have any knowledge of the readers so it doesn't have anything to shutdown or to flush.  

Which is created first has many implications on control flow (somewhat), and what methods need to be present on which parts.  In the former, the Reader will exist before the MeterProvider so the reader will need to have a MeterProvier, or something more specific, registered with it to be able to fetch the appropriate data.  In the latter, the MeterProvider will need some way to register Readers, or something more specific, for shutdown and flush.

In both cases, there is a need to be able to update the thing the Readers read with new instruments.  And if we are able to update the readers after the MeterProvider is created we will need some way for the Reader to update any existing instruments with its presence.

Reply:
I feel like I would like to discuss this more, but I'm realizing that having this conversation in this thread is not the appropriate place. I will comment more in https://github.com/open-telemetry/opentelemetry-go/issues/2809 about the reader design.

@MadVikingGod related to this PR, can we resolve this thread? Or, do we still need to remove the `Reader` part of the comment for now?

Reply:
I'm fine with leaving this and other parts of this PR in if we can review it and change it after we have confirmed the design.  I don't want to put these down now, make a design that finds something written here to be unnecessary, and then have to shoehorn in something that doesn't need to be there just because we wrote it down before.

Can we have this as asperations and not commitments?

Reply:
I am skeptical. One way or another the MeterProvider is going to know about its readers, and the specification does not call for supporting dynamic reader creation. There are still a lot of ways to factor the construction process.

Reply:
> I am skeptical. One way or another the MeterProvider is going to know about its readers, and the specification does not call for supporting dynamic reader creation. There are still a lot of ways to factor the construction process.

@jmacd I'm confused if you are directly concerned with the text "By default, the returned MeterProvider is configured with the default Resource and no Readers" or something else.

The default meter provider will not have any readers registered with it in either case. For the register-a-reader-with-meter-provider case, if you create a meter provider without any options (the default) it will not be configured with any readers. For the register-a-meter-provider-with-reader case, the default (and all other) meter provider will not be configured with any readers. I think this statement is appropriate for either design.

Reply:
I accept "and No readers".